_posts/2018-08-02-read-only-functions.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Serverless Security&#58; read-only functions with OpenFaaS
+description: Alex Ellis introduces the read-only functions feature for OpenFaaS - a security concept available with containers and Kubernetes to prevent tampering
+date: 2018-08-01
+image: /images/read-only/railings.jpg
+categories:
+  - security
+  - kubernetes
+  - swarm
+author_staff_member: alex
+---
+
+In this post I'll highlight one of the ways we're making OpenFaaS a more secure environment for your production functions and workloads.
+
+## What is a read-only function?
+
+All functions deployed to OpenFaaS are first built into immutable Docker images. We do this so that you get exactly the same results wherever you want to run your code. Read-only functions are functions which run with an enhanced security profile so that users cannot make changes to the underlying file-system in the container. This protects the underlying filesystem, shared libraries and the code for the function.
+
+We also use a read-only file-system in the OpenFaaS API Gateway which provides a REST API and user-friendly UI for your functions.
+
+## How does it work?
+
+In both Kubernetes and Docker Swarm, we enabled a feature in the built-in security context to make the root file-system `/` read-only. Before we applied this patch to the OpenFaaS providers we also changed all the OpenFaaS templates to run as non-root users to prevent tampering with system files, this change goes one step further and prevents a function from tampering with its own code.
```

Comment:
done

Reply:
For the BG image it would be ideal if we had a way to change the text colour - I have never seen this issue with my ghost blog.. wonder what it does differently? Maybe it's the way the text is overlaid?